src/shared/mount-util.c

```diff
@@ -475,67 +475,86 @@ int bind_remount_one_with_mountinfo(
         return 0;
 }
 
-int mount_move_root(const char *path) {
-        assert(path);
-
-        if (chdir(path) < 0)
-                return -errno;
-
-        if (mount(path, "/", NULL, MS_MOVE, NULL) < 0)
-                return -errno;
-
-        if (chroot(".") < 0)
-                return -errno;
-
-        return RET_NERRNO(chdir("/"));
-}
-
-int mount_pivot_root(const char *path) {
-        _cleanup_close_ int fd_oldroot = -EBADF, fd_newroot = -EBADF;
+int mount_switch_root(const char *path, MountAttrPropagationType type) {
+        _cleanup_close_ int fd_newroot = -EBADF;
+        unsigned int flags;
+        const char *prop_name;
 
         assert(path);
 
-        /* pivot_root() isn't currently supported in the initramfs. */
-        if (in_initrd())
-                return mount_move_root(path);
-
-        fd_oldroot = open("/", O_PATH|O_DIRECTORY|O_CLOEXEC|O_NOFOLLOW);
-        if (fd_oldroot < 0)
-                return log_debug_errno(errno, "Failed to open old rootfs");
+        switch (type) {
+        case PROPAGATION_INHERIT:
+                flags = 0;
+                prop_name = "inherited";
+                break;
+        case PROPAGATION_PRIVATE:
+                flags = MS_PRIVATE;
+                prop_name = "private";
+                break;
+        case PROPAGATION_DEPENDENT:
+                flags = MS_SLAVE;
+                prop_name = "dependent";
+                break;
+        case PROPAGATION_SHARED:
+                flags = MS_SHARED;
+                prop_name = "shared";
```

Comment:
so since enum-to-string mappings are super common, we have macros that generate safe translators for them, i.e. DEFINE_STRING_TABLE_LOOKUP() and stuff. I think we should do this for MountAttrPropagationType too, so that we have mount_attr_propagation_type_to_string(), then. This would be more inline with how we usually handle enums